File: pkg-r/tests/testthat/apps/tool-weather/app-06-tool-custom-output.R
library(shiny)
library(bslib)
library(ellmer)
library(shinychat)
library(weathR)

# This example replaces shinychat's default tool-result card with standalone UI.
# The activity row remains while the tool runs; when the result settles, the
# call leaves the row and the returned value box renders below it.

WeatherToolResult <- S7::new_class(
  "WeatherToolResult",
  parent = ContentToolResult,
  properties = list(
    location_name = S7::class_character
  )
)

contents_shinychat <- S7::new_external_generic(
  "shinychat",
  "contents_shinychat",
  "contents"
)

S7::method(contents_shinychat, WeatherToolResult) <- function(content) {
  current <- content@value[1, ]

  bslib::value_box(
    title = content@location_name,
    value = current$skies,
    showcase = bsicons::bs_icon("cloud-sun"),
    full_screen = TRUE,
    sprintf(
      "%s°F (High: %s°F, Low: %s°F)",
      current$temp,
      max(content@value$temp),
      min(content@value$temp)
    )
  )
}

get_weather_forecast <- tool(
  function(lat, lon, location_name) {
    WeatherToolResult(
      point_tomorrow(lat, lon, short = FALSE),
      location_name = location_name
    )
  },
  name = "get_weather_forecast",
  description = "Get the weather forecast for a location.",
  arguments = list(
    lat = type_number("Latitude"),
    lon = type_number("Longitude"),
    location_name = type_string("Name of the location for display to the user")
  ),
  annotations = tool_annotations(
    title = "Getting weather forecast",
    icon = bsicons::bs_icon("cloud-sun")
  )
)

ui <- function(req) {
  page_fillable(
    chat_ui("chat")
  )
}

server <- function(input, output, session) {
  client <- ellmer::chat("openai/gpt-4.1-nano")
  # client <- ellmer::chat_ollama(model = "mistral-nemo")
  client$register_tool(get_weather_forecast)
  chat_server("chat", client)
}

shinyApp(ui, server, enableBookmarking = "url")

File: pkg-py/tests/playwright/tools/weather/app_06_tool_custom_output.py
import faicons
import pandas as pd
from chatlas import ChatOpenAI, ContentToolResult
from shiny.express import ui
from shiny.ui import value_box
from shinychat import message_content_chunk
from shinychat.express import Chat
from shinychat.types import ChatMessage


class WeatherToolResult(ContentToolResult):
    """
    Custom tool result class for weather forecasts with custom value box output.
    This example shows how to use a custom tool result class that renders
    the weather data as a custom UI component (value box in R, custom HTML in Python).
    """

    location_name: str


@message_content_chunk.register
def _(message: WeatherToolResult):
    val = message.value
    high_temp = str(val["temperature"].max())
    low_temp = str(val["temperature"].min())
    current = val.iloc[0]

    content = value_box(
        message.location_name,
        str(current["temperature"]),
        f"{current['temperature']}°F (High: {high_temp}°F, Low: {low_temp}°F)",
        showcase=faicons.icon_svg("sun"),
        full_screen=True,
    )
    return ChatMessage(content=content)


def get_weather_forecast(
    lat: float, lon: float, location_name: str
) -> WeatherToolResult:
    """Get the weather forecast for a location."""
    # Mock detailed forecast data as a pandas DataFrame
    forecast_data = pd.DataFrame(
        {
            "time": ["Current", "06:00", "12:00", "18:00"],
            "temperature": [68, 65, 72, 66],
            "humidity": [75, 70, 65, 72],
            "conditions": ["Partly cloudy", "Clear", "Sunny", "Cloudy"],
            "wind_speed": [7, 5, 8, 6],
        }
    )
    return WeatherToolResult(value=forecast_data, location_name=location_name)


client = ChatOpenAI(model="gpt-4.1-nano")
client.register_tool(
    get_weather_forecast,
    annotations={"title": "Weather Forecast"},
)

ui.page_opts(title="Weather Tool - Custom Output")

chat = Chat(id="chat")
chat.ui()


@chat.on_user_submit
async def handle_user_input(user_input: str):
    response = await client.stream_async(user_input, content="all")
    await chat.append_message_stream(response)
